### ISSUE TITLE
Is `:a :b {}` the same triple as `:a :b true`

### ISSUE THREAD

**Issue by jeswr** [w3c/N3#185]
In https://github.com/rdfjs/types/issues/37 there has been a discussion around whether parsers should interpret `{}` as the literal `true`.

To me this seems reasonable. But I think an explicit decision needs to be made on this at a spec level, rather than it being implicit dependent on ones' interpretation of whether the last statement in [ground graph semantic](https://w3c.github.io/N3/spec/semantics.html#ground-graphs) is a syntactic or semantic one.

Once a decision is made I am happy to follow up with a PR containing either the positive or negative test case.

**Comment by doerthe** [comment]
The whole discussion here has several aspects.

First of all: In N3 and also in RDF the empty graph (already in N3's base interpretation and RDF's simple entailment) is always true. 

But in the examples you give, we do not just have the empty graph, we do have the empty graph used as a term and that is different.  If we have an N3 implication like

` {}=>{:birds :can :fly}.` 

We can translate the premise to an FOL-like truth, all together 

⊤ → can(birds,fly)

But: true is a boolean literal and not a formula, so the question here is rather whether we want to translate ⊤ to true. To add another example:

`:doerthe :denies {}.`

(freely translated to "Dörthe denies the truth" :)) then we have in FOL-style something like

denies(doerthe, ⊤)

Again, here the ⊤ stands for true as a formula and it is up to us whether we think that this is the same as the boolean literal true.

Currently we see them as equal, we state that true is syntactic sugar for {}. So, following your request, I would either make more explicit that we do that or define some IRI n3:trueGraphTerm instead. 

Could you (or others) think of any use cases in which we would like to be able to distinguish between true and {}? So far, I think it is a good solution to treat {} and true as equal, but I could be wrong.

**Comment by jeswr** [comment]
@william-vw ?

**Comment by william-vw** [comment]
> Could you (or others) think of any use cases in which we would like to be able to distinguish between true and {}? So far, I think it is a good solution to treat {} and true as equal, but I could be wrong.

Not really, no. Inversely, I can imagine use cases where one would want to differentiate between `false` and `{ }`, e.g., when a builtin fails (`false`) versus returning an empty formula (nothing was found / calculated). I can't think of an example where we'd have to distinguish between a `true` vs. `{ }` value, but maybe I'm lacking the imagination...

**Comment by doerthe** [comment]
I thought about it and think that it might be problematic that `{}` and `true` might get two raw types. 
Try the following with jen3 and EYE: https://n3-editor.herokuapp.com/n3/editor/s/jO8Pj1lC
We see that they differ and currently literals and graphs are disjoint (even though we do not explicitly state that in the semantics, but I guess we can get into trouble with some built-ins).

Also: `log:includes` only accepts graph terms as input, but `{}` is a literal according to EYE.

So maybe, it is not that easy after all...

**Comment by doerthe** [comment]
Ah, and I see that eye.js allows two types (but I guess you knew that :) )

**Comment by doerthe** [comment]
OK, so my main question is: can something be a formula and a literal at the same time? So far that is no problem, but I wonder whether it somehow clashes with RDF.

**Comment by sulivanShu** [comment]
"true is a boolean literal and not a formula" nor a graph.

True is a property of a formula or graph, but it is not the formula or the graph itself.

The fact a formula is true or false is a meta-theoretical property, not a theoretical value (even though we can write a formula that describes this fact).

"Could you (or others) think of any use cases in which we would like to be able to distinguish between true and {}?"

1. {} and "true"^^xsd:boolean are not of the same type. Confusion between {} and "true"^^xsd:boolean will cause an error whenever the difference in type is significant.

1. For a formula to be true or false, it must be asserted, that is, evaluated within a model. However, graph terms within a graph are not asserted.

1. `true`^^xsd:boolean, or `log:Truth`, are simply theoretical values ​​in the graph, not a meta-theoretical property of the graph term within a model. I can perfectly well write:

```
{:Person owl:disjointWith :Person} owl:hasValue "true"^^xsd:boolean .
```

And this is true in RDF/OWL, because the graph term is not asserted. No assertion = no evaluation in a model = no truth value. `owl:hasValue "true"^^xsd:boolean` is a predicate like any other, without particular meaning. Meaning come from valuation in a defined model.

Since these symbols denote different resources in RDF, OWL, XSD, or Log models, we will inevitably find use cases where their identity poses a problem.

**Comment by sulivanShu** [comment]
"true is a boolean literal and not a formula"

Indeed, and ⊤ is a formula, and not a boolean.  So ⊤ is not identical to true. ⊤ is a tautology: a formula true in all model (for example A = A).

And yes, in RDF, {} ≡⊤

because all RDF graph are true and empty graph is a valid graph in RDF.

These axioms allow us to solve the N3 formulas that involve the empty graph, for example:

Implication: `{}=>{:birds :can :fly}.`

Antecedent `{}` # because "The empty graph is simply entailed by any graph" ([source](https://w3c.github.io/rdf-semantics/spec/ ))

Consequent: `:birds :can :fly .`

**Comment by doerthe** [comment]
@sulivanShu So, I read that as you agreeing with me: 
There should be a difference between `{}` and `"true"^^xsd:boolean`. 
Do I get you right?

**Comment by josd** [comment]
It came after we accepted using `premise => false` to express inference fuses.

**Comment by sulivanShu** [comment]
> [@sulivanShu](https://github.com/sulivanShu) So, I read that as you agreeing with me: There should be a difference between `{}` and `"true"^^xsd:boolean`. Do I get you right?

Yes. A formula should not be confused with its truth value.

> It came after we accepted using `premise => false` to express inference fuses.

This is problematic. In the context of this discussion:

"⊤" is the symbol for "tautology," not "true."

"⊥" is the symbol for "absurdity," not "false."

"⊤" and "⊥" are formulas, not truth values.

The implication "=>" is a relation that only accepts formulas (even the empty formulas) as arguments, not Boolean values.

This is a type error for this relation.

At the same time, I understand the error, because the meaning of the symbols "⊤" or "⊥" is not fixed and depends on the context.

Wikipedia details the polysemy of these symbols:

> The tee (⊤, \top in LaTeX), also called down tack (as opposed to the up tack) or verum,[1] is a symbol used to represent: 

> The truth value of being true in logic, or a sentence (e.g., formula in propositional calculus) which is unconditionally true.[2][3][4] By definition, every tautology is logically equivalent to the verum.

[source](https://en.wikipedia.org/wiki/Tee_(symbol))

> "Up tack" is the Unicode name for a symbol (⊥, \bot in LaTeX, U+22A5 in Unicode[1]) that is also called "bottom",[2][3] "falsum",[4] "absurdum",[5] or "absurdity",[6][7][3] depending on context. It is used to represent: 

> The truth value 'false', or a logical constant denoting a proposition in logic that is always false. (The names (falsum, absurdum, and absurdity) come from this context.

[source](https://en.wikipedia.org/wiki/Up_tack)

But a truth value and a formula that is unconditionally true or false are two completely different things.

Thus, it would be better to use IRI like `log:verum` or `log:falsum`, and not `"true"^^xsd:boolean` or `"false"^^xsd:boolean`.

**Comment by josd** [comment]
Let us make conformance tests:

```
@prefix log: <http://www.w3.org/2000/10/swap/log#>.
@prefix : <http://example.org/>.

{
  {} log:notEqualTo true .
}
=>
{
  :test :is true .
} .
```
should give `:test :is true .`

It works fine with the eyeling reasoner

https://eyereasoner.github.io/eyeling/playground?state=g.H4sIAAAAAAAACkWMQQrDIBBFrzJMt0Ftu5M2dNMbdOnGxcQGjFqdYCB499KQkt2H995fkVDjI2UaxgV8dBpub-akpay1inoVMTt5UUrJs5Kl2iR9dKdemPCPjoIWOyVPW_IzTFhNAFjbdhwiPz-z9a8InGcCYUIz4d7vlmYqDHosBwWBHWbUg_WFOpz21b4mHiAVtQAAAA

It does however not work fine with the eye reasoner 👎 

@phochste can you add it to https://codeberg.org/phochste/notation3tests

**Comment by TallTed** [comment]
Coming late to this discussion, and probably not fully understanding the logical aspects, I noticed an apparent equivalency, in conversation circa June-July 2023, which may not actually be equivalent. That is, `{}` (empty) and `{ }` (enclosed space). I have hopes that this has no implication to the subsequent discussion, but it seemed worth flagging, _just in case_.

**Comment by sulivanShu** [comment]
Most programming languages distinguish between meaningful whitespace, which is typically used to separate terminals, and meaningless whitespace, which is used solely for formatting and readability.

While the tokenizer interprets meaningful whitespace, it ignores meaningless whitespace because it carries no semantic content.

Thus, `{ }`, `{  }`, or `{	}` are equivalent to `{}` because the whitespace they contain is meaningless.

The meaningful whitespace in Turtle is defined [here](https://www.w3.org/TR/rdf12-turtle/#sec-grammar-ws). Other whitespace should be considered meaningless, not syntax errors.

**Comment by TallTed** [comment]
@sulivanShu — Yes, I am accustomed to the variance in interpretation of meaningful and meaningless whitespace. I have spent too many hours fixing Oracle configuration files where line-starting whitespace was syntactically important, especially but not only spaces-vs-tabs and mixes thereof. That is why I raised my concern about the `{}` and `{ }` seen earlier in this issue. I'm glad to hear the whitespace doesn't matter in this case.
